src/nitypes/vector/_vector.py

```diff
@@ -0,0 +1,148 @@
+from __future__ import annotations
+
+from typing import Any, Generic, Union
+
+from typing_extensions import TypeVar, final
+
+from nitypes._exceptions import invalid_arg_type, invalid_arg_value
+from nitypes.waveform._extended_properties import (
+    UNIT_DESCRIPTION,
+    ExtendedPropertyDictionary,
+)
+
+_VectorType = TypeVar("_VectorType", bound=Union[bool, int, float, str])
+
+
+@final
+class Vector(Generic[_VectorType]):
+    """A Vector of scalar values with units information.
+
+    Constructing
+    ^^^^^^^^^^^^
+
+    To construct a vector data object, use the :class:`Vector` class:
+
+    >>> Vector([False, True])
+    nitypes.vector.Vector(values=[False, True], units='')
+    >>> Vector([0, 1, 2])
+    nitypes.vector.Vector(values=[0, 1, 2], units='')
+    >>> Vector([5.0, 6.0], 'volts')
+    nitypes.vector.Vector(values=[5.0, 6.0], units='volts')
+    >>> Vector(["one", "two"], "volts")
+    nitypes.vector.Vector(values=['one', 'two'], units='volts')
+    """
+
+    __slots__ = [
+        "_values",
+        "_extended_properties",
+    ]
+
+    _values: list[_VectorType]
+    _extended_properties: ExtendedPropertyDictionary
+
+    def __init__(
```

Comment:
Suggestion: do it like numpy but with types instead of dtypes
 
```
v1 = Vector([True])
reveal_type(v1)  # Vector[bool]
v2 = Vector([], type=bool)
reveal_type(v2)  # Vector[bool]
```

Note that if you shadow the name `type`, you can still access it via `builtins.type`